### PR TITLE
AutoYaST: Support to configure the network during the first stage by default

### PR DESCRIPTION
**Pull request by teclator**
This PR is related to https://github.com/yast/yast-autoinstallation/pull/606 where the problem and solutions are already described, so, use it as a base context.

## Problem (specific to this module)

There are some issues we should fix in order to write the configuration during the first stage.

**1. Udev rules are not refreshed after writting**

The InterfacesWriter did not refreshed the written udev rules in case of an autoinstallation as we can see here:

https://github.com/yast/yast-network/pull/1059/files#diff-48ec4731f89821c19dc203da64dc00a5L108

But during the first stage, the imported network configuration can be applied before the proposal, and, at that point, a rename of the interfaces is needed. Thus, we should distinguish between being at the end of the installation (already chrooted) or not.

By now, we have added a parameter which determines whether the writer should also reload the configuration or not.

**2. Udev rules are written to the inst-sys and need to be copied to the target system**

The `udev_persistent` agent  or in general `non y2 agents` do not permit to change the agent root location

https://yastgithubio.readthedocs.io/en/latest/development-tips/#changing-the-root-location-for-agents

in case that we are at the end of the instalation, the files inst-sys files are copied before written. If the config is written later as it is, at least the udev rules should be copied.

**3. Linuxrc and imported configuration is somehow merge even when keep_install_network is failse**

The autoinst reader uses the current sysconfig as the base one

https://github.com/yast/yast-network/pull/1059/commits/f4abc6a22e469bb6f690df5eb1bd3acf87fc6087#diff-6dc211a0076a98b7941e86c01765b477L51

Applying the changes imported over it. But only interfaces are needed in order to update properly the associated udev rules, and also for removing the ones not imported, specially if it is selected to not keep the linuxrc configured network.

**4. The NetworkAutoyast checks the profile directly and writes the config based on the presence of the second_stage**

If the idea is to remove the networking section during autosetup, the networking section will not be present when calling it at the end of the installation. Thus, the module need to be updated. But also, it checks whether the second stage is present or not in order to write the config.

The config should not be written again in case it was already configured before the proposal and the check of the second stage does not make sense anymore
/761872905d0b162fc5f46329398763f1116d69d8#diff-24ba5c89e21a968be8706d02f59adb43L314

## Solution

- The refresh of of the udev rules respect now the Lan.write_only variable, thus, during the setup_before_proposal, them should be refreshed, but in case of the end of the installation them should not
- If the imported network configuration is written at the end of the installation, the udev rules are copied to the target system after being modified.
- The autoinst reader will only use the current system interfaces apart of the imported configuration as the internal yast_config. In case of merged, it will be already defined by the merged profile structure.
- Some logic have been removed from NetworkAutoyast, and now will mainly write the imported config.
- Moved the auto client to its own class under y2network namespace improving the unit test coverage.

## Tests

- Added some uni tests and do some tests manually, in this area, and specially with udev rules handling, integration tests are very important.

### CODE REVIEW

File: src/lib/y2network/sysconfig/interfaces_writer.rb [hunk 1]
```diff
@@ -33,7 +31,12 @@ module Sysconfig
     # hardware specific configuration through udev rules.
     #
     # @see Y2Network::InterfacesCollection
+    # @param reload [Boolean] whether the udev rules should be reloaded or not
```
Comment (imobachgs) [inline src/lib/y2network/sysconfig/interfaces_writer.rb:34]: Please, move this line to its correct place :wink: (1 line below).

File: src/modules/Lan.rb [hunk 2]
```diff
@@ -106,6 +107,11 @@ def main
       @configs = {}
     end
 
+    # @return [Boolean]
+    attr_accessor :write_only
+    # @return [Autoinst::Config]
+    attr_writer :autoinst
```
Comment (imobachgs) [inline src/modules/Lan.rb:113]: :+1:

File: src/lib/network/network_autoyast.rb [hunk 3]
```diff
@@ -101,22 +101,24 @@ def set_network_service
...
-      end
-
-      configure_submodule(Lan, ay_configuration, write: write)
+      return false if Lan.autoinst.before_proposal
+
+      # force a write only as it is run at the end of the installation and in
+      # the target system
+      original_value = Lan.write_only
+      Lan.write_only = true
+      ret = Lan.Write
```
[the remark this answers is not in the document]
Comment (teclator) [reply src/lib/network/network_autoyast.rb:119]: Yep, already thinking on adding that.

File: src/lib/y2network/clients/auto.rb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    private
+
+      def merge_current_config?
+        !!Yast::Lan.autoinst.keep_install_network
+      end
+
+      # Convert data from native network to autoyast for XML
+      # @param [Hash] settings native network settings
+      # @return [Hash] autoyast network settings
+      def adapt_for_autoyast(settings)
```
[the remark this answers is not in the document]
Comment (teclator) [reply src/lib/y2network/clients/auto.rb:121]: Completely agree. So far, have just moved the code to a class and added unit tests, but this is generally the ToAY we wanted to get rid of.

**Pull request merged**